### PR TITLE
[GlobalISel][AArch64] Legalize G_ADD, G_SUB, G_AND, G_OR, and G_XOR for SVE

### PR DESCRIPTION
**Pull request by tschuett**
Credits: https://github.com/llvm/llvm-project/pull/72976

LLVM ERROR: cannot select: %3:zpr(<vscale x 2 x s64>) = G_MUL %0:fpr, %1:fpr (in function: xmulnxv2i64)

;; mul
define void @xmulnxv2i64(<vscale x 2 x i64> %a, <vscale x 2 x i64> %b, ptr %p) { 
entry:
  %c = mul <vscale x 2 x i64> %a, %b
  store <vscale x 2 x i64> %c, ptr %p, align 16
   ret void
}

define void @mulnxv4i32(<vscale x 4 x i32> %a, <vscale x 4 x i32> %b, ptr %p) { 
entry:
  %c = mul <vscale x 4 x i32> %a, %b
  store <vscale x 4 x i32> %c, ptr %p, align 16
   ret void
}

define void @mulnxv8i16(<vscale x 8 x i16> %a, <vscale x 8 x i16> %b, ptr %p) { 
entry:
  %c = mul <vscale x 8 x i16> %a, %b
  store <vscale x 8 x i16> %c, ptr %p, align 16
  ret void
}

define void @mulnxv16i8(<vscale x 16 x i8> %a, <vscale x 16 x i8> %b, ptr %p) { 
entry:
  %c = mul <vscale x 16 x i8> %a, %b
  store <vscale x 16 x i8> %c, ptr %p, align 16
  ret void
}

### CODE REVIEW

**Comment by tschuett** [comment]
How is GlobalISel going to handle predication, e.g., `ABS`?

**Comment by tschuett** [comment]
First bet, there is no pattern for mul for SVE.

**Comment by tschuett** [comment]
```
(base) llvm-project2 % ag "\(ADD_ZZZ_H" lib/Target/AArch64/AArch64GenGlobalISel.inc
8233:      // (add:{ *:[nxv8i16] } nxv8i16:{ *:[nxv8i16] }:$Op1, nxv8i16:{ *:[nxv8i16] }:$Op2)  =>  (ADD_ZZZ_H:{ *:[nxv8i16] } ?:{ *:[nxv8i16] }:$Op1, ?:{ *:[nxv8i16] }:$Op2)
(base)  llvm-project2 % ag "\(SUB_ZZZ_H" lib/Target/AArch64/AArch64GenGlobalISel.inc
10643:      // (sub:{ *:[nxv8i16] } nxv8i16:{ *:[nxv8i16] }:$Op1, nxv8i16:{ *:[nxv8i16] }:$Op2)  =>  (SUB_ZZZ_H:{ *:[nxv8i16] } ?:{ *:[nxv8i16] }:$Op1, ?:{ *:[nxv8i16] }:$Op2)
(base) llvm-project2 % ag "\(MUL_ZZZ_H" lib/Target/AArch64/AArch64GenGlobalISel.inc
```

There indeed patterns missing for the SVE G_MUL.

**Comment by tschuett** [comment]
success:
```
> ag "\(AND_ZZZ" lib/Target/AArch64/AArch64GenGlobalISel.inc
> ag "\(EOR_ZZZ" lib/Target/AArch64/AArch64GenGlobalISel.inc
> ag "\(ORR_ZZZ" lib/Target/AArch64/AArch64GenGlobalISel.inc
```

**Comment by tschuett** [comment]
```
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:10288:
llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:3491:21: note: instantiated from multiclass
  defm MUL_ZZZ    : sve2_int_mul<0b000,  "mul",   AArch64mul>;
```

```
>  ./bin/llvm-tblgen -gen-global-isel -warn-on-skipped-patterns 
```

**Comment by tschuett** [comment]
The  **complete** error message about importing MUL_ZZZ:
```
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:1035:
llvm/lib/Target/AArch64/SVEInstrFormats.td:3694:3: warning: Skipped pattern: Pattern operator lacks an equivalent Instruction (AArch64ISD::MUL_PRED)
  def : SVE_2_Op_Pat<nxv16i8, op, nxv16i8, nxv16i8, !cast<Instruction>(NAME # _B)>;
  ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:10288:
llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:3491:21: note: instantiated from multiclass
  defm MUL_ZZZ    : sve2_int_mul<0b000,  "mul",   AArch64mul>;
                    ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:1035:
llvm/lib/Target/AArch64/SVEInstrFormats.td:3695:3: warning: Skipped pattern: Pattern operator lacks an equivalent Instruction (AArch64ISD::MUL_PRED)
  def : SVE_2_Op_Pat<nxv8i16, op, nxv8i16, nxv8i16, !cast<Instruction>(NAME # _H)>;
  ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:10288:
llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:3491:21: note: instantiated from multiclass
  defm MUL_ZZZ    : sve2_int_mul<0b000,  "mul",   AArch64mul>;
                    ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:1035:
llvm/lib/Target/AArch64/SVEInstrFormats.td:3696:3: warning: Skipped pattern: Pattern operator lacks an equivalent Instruction (AArch64ISD::MUL_PRED)
  def : SVE_2_Op_Pat<nxv4i32, op, nxv4i32, nxv4i32, !cast<Instruction>(NAME # _S)>;
  ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:10288:
llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:3491:21: note: instantiated from multiclass
  defm MUL_ZZZ    : sve2_int_mul<0b000,  "mul",   AArch64mul>;
                    ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:1035:
llvm/lib/Target/AArch64/SVEInstrFormats.td:3697:3: warning: Skipped pattern: Pattern operator lacks an equivalent Instruction (AArch64ISD::MUL_PRED)
  def : SVE_2_Op_Pat<nxv2i64, op, nxv2i64, nxv2i64, !cast<Instruction>(NAME # _D)>;
  ^
Included from llvm/lib/Target/AArch64/AArch64.td:37:
Included from llvm/lib/Target/AArch64/AArch64InstrInfo.td:10288:
llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:3491:21: note: instantiated from multiclass
  defm MUL_ZZZ    : sve2_int_mul<0b000,  "mul",   AArch64mul>;
```

**Comment by tschuett** [comment]
There are several occurrences.
```
def AArch64mul_p  : SDNode<"AArch64ISD::MUL_PRED",  SDT_AArch64Arith>;
```
where AArch64ISD::MUL_PRED is C++ and not tablegen.

**Comment by tschuett** [comment]
Ping.

**Comment by davemgreen** [comment]
For the SVE issue you are seeing - some instructions in SVE are "unpredicated", many require a predicate even if it is an all-on predicate vector. mul I believe is one where there is an predicated +sve instruction, and an unpredicated +sve2 version. SDAG will convert them into MUL_PRED, but that has obvious downsides.

Do you mind if we have a first patch that just gets the arg-passing basics out of the way first, and start adding instruction support separately? Or does that all work already and all of this is for instruction support? We should be able to add tests that pass in values and store them, or load values and return them. I feel we need to make sure SVE is decently designed. I suppose that without any other suggestions we can follow how SDAG behaves.

**Comment by tschuett** [comment]
At your speed.

**Comment by tschuett** [comment]
My plan was to legalize more instructions, but eventually we will run into predication and/or import issues.

**Comment by tschuett** [comment]
[Translate legal SVE formal arguments and select COPY for SVE](https://github.com/llvm/llvm-project/pull/95236)

Argument passing is already implemented.

**Comment by tschuett** [comment]
Ping.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 1]
```diff
@@ -127,7 +128,37 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .clampNumElements(0, v2s64, v2s64)
       .moreElementsToNextPow2(0);
 
-  getActionDefinitionsBuilder({G_ADD, G_SUB, G_MUL, G_AND, G_OR, G_XOR})
+  auto &IntegerArithmeticActions =
+      getActionDefinitionsBuilder({G_ADD, G_SUB, G_AND, G_OR, G_XOR});
+  if (HasSVE)
+    IntegerArithmeticActions.legalFor({nxv16s8, nxv8s16, nxv4s32, nxv2s64});
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:134]: Can you rebase now that a3010c77910c706be4c51ce is in?

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h [hunk 2]
```diff
@@ -1198,6 +1197,11 @@ class LegalizeRuleSet {
     assert(MinTy.getElementType() == MaxTy.getElementType() &&
            "Expected element types to agree");
 
+    if (MinTy.isScalableVector())
+      return actionIf(LegalizeAction::Unsupported, always);
+    if (MaxTy.isScalableVector())
+      return actionIf(LegalizeAction::Unsupported, always);
```
Comment (davemgreen) [inline llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h:1203]: This isn't used/needed yet? It should probably just be an assert with the current interfaces. We will need to find a way to fix those in the future.

File: llvm/test/CodeGen/AArch64/sve-integer.ll [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+define void @addnxv2i64(<vscale x 2 x i64> %a, <vscale x 2 x i64> %b, ptr %p) {
+; CHECK-LABEL: addnxv2i64:
+; CHECK:       // %bb.0: // %entry
+; CHECK-NEXT:    add z0.d, z0.d, z1.d
+; CHECK-NEXT:    ptrue p0.d
+; CHECK-NEXT:    st1d { z0.d }, p0, [x0]
+; CHECK-NEXT:    ret
+entry:
+  %c = add <vscale x 2 x i64> %a, %b
+  store <vscale x 2 x i64> %c, ptr %p, align 16
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/sve-integer.ll:15]: Can these all return directly, as opposed to storing?
Can we reuse an existing test file like llvm/test/CodeGen/AArch64/sve-int-arith.ll?

**Comment by tschuett** [comment]
Thanks for taking a look.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:134]: It should be
```
getActionDefinitionsBuilder({G_ADD, G_SUB, G_MUL, G_AND, G_OR, G_XOR})
    .legalFor({s32, s64, v2s32, v2s64, v4s32, v4s16, v8s16, v16s8, v8s8})
    .legalFor(HasSVE, {nxv16s8, nxv8s16, nxv4s32, nxv2s64})
    ....
```
(You could make the argument that the HasSVE is unnecessary, as scalable-vectors are not supported without them and will just hit some failure elsewhere anyway. We might as well add it in though, now that we can without messing up the Definitions).

File: llvm/test/CodeGen/AArch64/sve-integer.ll (hunk 3 above)
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/sve-integer.ll]: Can we use one of the existing tests with a new run line?

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/sve-integer.ll]: sve-int-arith.ll has too much illegal resp. unsupported tests.

Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/sve-integer.ll]: I see, it is still crashing trying to lower the arguments, as opposed to falling back. Can you move this into AArch64/GlobalISel/sve-integer.ll for the moment then, and we can try to de-duplicate them later on.

File: llvm/test/CodeGen/AArch64/sve-integer.ll [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+define <vscale x 16 x i8> @xornxv16i8(<vscale x 16 x i8> %a, <vscale x 16 x i8> %b) {
+; CHECK-LABEL: xornxv16i8:
+; CHECK:       // %bb.0: // %entry
+; CHECK-NEXT:    eor z0.d, z0.d, z1.d
+; CHECK-NEXT:    ret
+entry:
+  %c = xor <vscale x 16 x i8> %a, %b
+  ret <vscale x 16 x i8> %c
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/sve-integer.ll:207]: We `ret` and it works fine.

File: llvm/test/CodeGen/AArch64/sve-integer.ll (hunk 3 above)
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/sve-integer.ll]: sve-integer.ll works with `ret`. Long term, we should merge the two files, but our SVE support is really limited right now.

Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/sve-integer.ll]: Yep this is no longer about ret - the tests look good now. This is about the duplication and moving them to the GlobalISel directory for the time being until more stuff works without crashing.

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**